docs/structured-streaming-programming-guide.md

```diff
@@ -0,0 +1,888 @@
+df.groupBy(
+     "type",
+     window(time, "1 minute"))
+  .avg(df.col("signal"))
+{% endhighlight %}
+
+</div>
+<div data-lang="java"  markdown="1">
+
+</div>
+<div data-lang="python"  markdown="1">
+
+</div>
+</div>
+
+This windowing based on group automatically handles late arriving data. 
+
+![Handling Late Data](img/structured-streaming-late-data.png)
+
+### Joins Operations
+Streaming DataFrames can be joined with static DataFrames to create new streaming DataFrames. Here are a few examples.
+
+<div class="codetabs">
+<div data-lang="scala"  markdown="1">
+
+{% highlight scala %}
+val staticDf = spark.read. … 
+val streamingDf = spark.readStream. … 
+
+streamingDf.join(staticDf, “type”)                     // inner equi-join with a static DF
+streamingDf.join(staticDf, “type”, “right_join”)  // right outer join with a static DF  
+
+{% endhighlight %}
+
+</div>
+<div data-lang="java"  markdown="1">
+
+</div>
+<div data-lang="python"  markdown="1">
+
+</div>
+</div>
+
+### Unsupported Operations
+However, note that all of the operations applicable on static DataFrames/Datasets are not supported in streaming DataFrames/Datasets yet. While some of these unsupported operations will be supported in future releases of Spark, there are others which are fundamentally hard to implement on streaming data efficiently. As of Spark 2.0, some of the unsupported operations are as follows
+
+- Multiple aggregations (i.e. a chain of aggregations on a streaming DF) are not yet supported
+
+- Limit and take first N rows are not supported
+
+- Distinct and sorting operations are not supported
+
+- Stream-batch outer joins are conditionally supported
+
+    + Full outer join not allowed
+
+    + Left outer join with a streaming DF on the left is not supported
+
+    + Right outer join with a streaming DF on the right is not supported
+
+- Stream-stream joins are not yet supported
+
+In addition, there are some Dataset methods that will not work on streaming Datasets. They are actions that will immediately run queries and return results, which does not makes sense on a streaming Dataset. Rather those functionalities can be done by explicitly starting a streaming query (see the next section regarding that).
+
+- `count()` - Cannot return a single count from a streaming Dataset. Instead, use `ds.groupBy.count()` which returns a streaming Dataset containing a running count. 
+
+- `foreach()` - Instead use `ds.writeStream.foreach(...)` (see next section).
+
+- `show()` - Instead use the console sink (see next section).
+
+If you try any of these operations, you will see an AnalysisException like "operation XYZ is not supported with streaming DataFrames/Datasets".
+
+## Starting Streaming Queries
+Once you have defined the final result DataFrame/Dataset, all that is left is for you start the StreamingQuery. To do that, you have to use the `DataStreamWriter`, through you have to specify the following. 
```

Comment:
through you have to specify => through which you have to specify
(I'm assuming)